sql/schemachanger: support ALTER POLICY DDL

spilchen opened the pull request:
Previously, attempts to execute ALTER POLICY DDL statements failed with an unimplemented error.

This change enables support for ALTER POLICY by leveraging the existing policy-related scpb elements, which already accounted for alteration scenarios.

Closes #136996
Closes #136997

Epic: CRDB-11724
Release note: none

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/143356)
<!-- Reviewable:end -->

rafiss reviewed (approved):
lgtm! i just had optional suggestions, but that's nice to see that the implementation itself could build upon the previous work without needing any new abstractions.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/143356#-:-OM7k_rf5YcTzf5U-134:b3za9qy)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @spilchen)

___
*[`pkg/sql/schemachanger/scbuild/internal/scbuildstmt/alter_policy.go` line 26 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/143356#-OM7iUbx4a4VMmlarek3:-OM7iUbx4a4VMmlarek4:b-r62ml7) ([raw file](https://github.com/cockroachdb/cockroach/blob/354af8c176896d56eaab6f7f51b16bce3a843ce4/pkg/sql/schemachanger/scbuild/internal/scbuildstmt/alter_policy.go#L26)):*
> ```Go
> 	// Alter of a policy implies that it must exist.
> 	policyElems := b.ResolvePolicy(tbl.TableID, n.PolicyName, ResolveParams{})
> 	policy := policyElems.FilterPolicy().MustGetZeroOrOneElement()
> ```

if the policy must exist here, would it make sense to return an error if `policy` is nil or use `MustGetOneElement`?
___
*[`pkg/sql/logictest/testdata/logic_test/row_level_security` line 1372 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/143356#-OM7jF2l2UEmmaEKY0kL:-OM7jF2l2UEmmaEKY0kM:b-2k6asj) ([raw file](https://github.com/cockroachdb/cockroach/blob/354af8c176896d56eaab6f7f51b16bce3a843ce4/pkg/sql/logictest/testdata/logic_test/row_level_security#L1372)):*
> ```
>                     CREATE POLICY p_sel ON public.alter_policy_table AS PERMISSIVE FOR SELECT TO aux1, alter_policy_role, aux2 USING (c1 != 1:::INT8)
> 
> # Spaces in the expressions mess up the logic tests column count, so we replace them with underscores.
> ```

i believe we can avoid the whitespace replacement if the test uses `ORDER BY` in the query rather than using the `rowsort` option. see: https://github.com/cockroachdb/cockroach/blob/6c2c34311ca8cfa0370006832a4133782431acd6/pkg/sql/logictest/logic.go#L2955-L2970


<!-- Sent from Reviewable.io -->

spilchen committed:
sql/schemachanger: support ALTER POLICY DDL

Previously, attempts to execute ALTER POLICY DDL statements failed with
an unimplemented error.

This change enables support for ALTER POLICY by leveraging the existing
policy-related scpb elements, which already accounted for alteration
scenarios.

Closes #136996
Closes #136997

Epic: CRDB-11724
Release note: none

spilchen reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/143356)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @rafiss)

___
*[`pkg/sql/schemachanger/scbuild/internal/scbuildstmt/alter_policy.go` line 26 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/143356#-OM7iUbx4a4VMmlarek3:-OM843XL8sdG8vlAqIAR:b-gqp850) ([raw file](https://github.com/cockroachdb/cockroach/blob/354af8c176896d56eaab6f7f51b16bce3a843ce4/pkg/sql/schemachanger/scbuild/internal/scbuildstmt/alter_policy.go#L26)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

if the policy must exist here, would it make sense to return an error if `policy` is nil or use `MustGetOneElement`?
</blockquote></details>

Makes sense. I'll use `MustGetOneElement`.
___
*[`pkg/sql/logictest/testdata/logic_test/row_level_security` line 1372 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/143356#-OM7jF2l2UEmmaEKY0kL:-OM84CTf3khFBN530PLf:bqz0pa4) ([raw file](https://github.com/cockroachdb/cockroach/blob/354af8c176896d56eaab6f7f51b16bce3a843ce4/pkg/sql/logictest/testdata/logic_test/row_level_security#L1372)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

i believe we can avoid the whitespace replacement if the test uses `ORDER BY` in the query rather than using the `rowsort` option. see: https://github.com/cockroachdb/cockroach/blob/6c2c34311ca8cfa0370006832a4133782431acd6/pkg/sql/logictest/logic.go#L2955-L2970
</blockquote></details>

Ah, good to know — this one had me scratching my head.


<!-- Sent from Reviewable.io -->

spilchen commented:
TFTR!

bors r+
